explore/exploration.h

```diff
@@ -0,0 +1,207 @@
+      OutputIt pdf_new_last = pdf_first + ((std::min)(num_actions_scores, num_actions_pdf));
+
+      // zero out pdf
+      for (OutputIt d = pdf_new_last; d != pdf_last; ++d)
+        *d = 0;
+
+        pdf_last = pdf_new_last;
+    }
+
+    if (num_actions_scores == 0)
+      return S_EXPLORATION_OK;
+
+    float norm = 0.;
+    float max_score = *std::max_element(scores_begin, scores_last);
+
+    InputIt s = scores_begin;
+    for (OutputIt d = pdf_first; d != pdf_last && s != scores_last; ++d, ++s)
+    {
+      float prob = exp(lambda*(*s - max_score));
+      norm += prob;
+
+      *d = prob;
+    }
+
+    // normalize
+    for (OutputIt d = pdf_first; d != pdf_last; ++d)
+      *d /= norm;
+
+    return S_EXPLORATION_OK;
+  }
+
+  template<typename InputIt, typename OutputIt>
+  int generate_softmax(float lambda, InputIt scores_begin, InputIt scores_last, OutputIt pdf_first, OutputIt pdf_last)
+  {
+    typedef typename std::iterator_traits<InputIt>::iterator_category scores_category;
+    typedef typename std::iterator_traits<OutputIt>::iterator_category pdf_category;
+
+    return generate_softmax(lambda, scores_begin, scores_last, scores_category(), pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename InputIt, typename OutputIt>
+  int generate_bag(InputIt top_actions_begin, InputIt top_actions_last, std::input_iterator_tag top_actions_tag, OutputIt pdf_first, OutputIt pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    if (pdf_first == pdf_last || pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
+
+    uint32_t num_models = std::accumulate(top_actions_begin, top_actions_last, 0);
+    if (num_models == 0)
+    {
+      if (pdf_last - pdf_first > 0)
+      {
+        *pdf_first = 1;
+        for (OutputIt d = pdf_first + 1; d != pdf_last; ++d)
+          *d = 0;
+      }
+
+      return S_EXPLORATION_OK;
+    }
+
+    // divide late to improve numeric stability
+    InputIt t_a = top_actions_begin;
+    for (OutputIt d = pdf_first; d != pdf_last && t_a != top_actions_last; ++d, ++t_a)
+      *d = *t_a / (float)num_models;
+
+    return S_EXPLORATION_OK;
+  }
+
+  template<typename InputIt, typename OutputIt>
+  int generate_bag(InputIt top_actions_begin, InputIt top_actions_last, OutputIt pdf_first, OutputIt pdf_last)
+  {
+    typedef typename std::iterator_traits<InputIt>::iterator_category top_actions_category;
+    typedef typename std::iterator_traits<OutputIt>::iterator_category pdf_category;
+
+    return generate_bag(top_actions_begin, top_actions_last, top_actions_category(), pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename It>
+  int enforce_minimum_probability(float min_prob, bool update_zero_elements, It pdf_first, It pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    if (pdf_last < pdf_first)
+      return E_EXPLORATION_BAD_RANGE;
+
+	  size_t num_actions = pdf_last - pdf_first;
+
+    if (min_prob > 0.999) // uniform exploration
```

Comment:
This is still the wrong guard.   You want 1 / support_size.

Reply:
Ah, I see. This isn't the minimum probability per action.  It's actions * minimum probability per action.